python-package/lightgbm/basic.py

```diff
@@ -843,6 +827,290 @@ def num_feature(self):
+
+    def create_valid(self, data, label=None, weight=None, group=None,
+                     silent=False, params=None):
+        """
+        Create validation data align with current dataset
+
+        Parameters
+        ----------
+        data : string/numpy array/scipy.sparse
+            Data source of _InnerDataset.
+            When data type is string, it represents the path of txt file
+        label : list or numpy 1-D array, optional
+            Label of the training data.
+        weight : list or numpy 1-D array , optional
+            Weight for each instance.
+        group : list or numpy 1-D array , optional
+            Group/query size for dataset
+        silent : boolean, optional
+            Whether print messages during construction
+        params: dict, optional
+            Other parameters
+        """
+        return Dataset(data, label=label, max_bin=self.max_bin, reference=self,
+                       weight=weight, group=group, predictor=self.predictor,
+                       silent=silent, params=params, free_raw_data=self.free_raw_data,
+                       lazy_init=self.lazy_init)
+
+    def construct(self):
+        """Lazy init"""
+        if self.inner_dataset is None:
+            if self.reference is not None:
+                if self.used_indices is None:
+                    self.inner_dataset = self.reference._get_inner_dataset().create_valid(
+                        self.data, self.label,
+                        self.weight, self.group,
+                        self.silent, self.params)
+                else:
+                    """construct subset"""
+                    self.inner_dataset = self.reference._get_inner_dataset().subset(
+                        self.used_indices, self.params)
+            else:
+                self.inner_dataset = _InnerDataset(self.data, self.label, self.max_bin,
+                    None, self.weight, self.group, self.predictor,
+                    self.silent, self.feature_name, self.categorical_feature, self.params)
+            if self.free_raw_data:
+                self.data = None
+
+    def _get_inner_dataset(self):
+        self.construct()
+        return self.inner_dataset
+
+    def __is_constructed(self):
+        return self.inner_dataset is not None
+
+    def set_categorical_feature(self, categorical_feature):
+        if self.categorical_feature == categorical_feature:
+            return
+        if self.data is not None:
+            self.categorical_feature = categorical_feature
+            self.inner_dataset = None
+        else:
+            raise Exception("Cannot set categorical feature after freed raw data")
+
+    def set_predictor(self, predictor):
+        if predictor is self.predictor:
+            return
+        if self.data is not None:
+            self.predictor = predictor
+            self.inner_dataset = None
+        else:
+            raise Exception("Cannot set predictor after freed raw data")
```

Comment:
I see. I think you can give some hints about setting free_raw_data to False in that exception message.